### ISSUE TITLE
Some issues and wrinkles with telemetry metrics

### ISSUE THREAD

**Issue by ancazamfir** [informalsystems/ibc-rs#2408]
## Summary of Bug
I was trying to make sense of the telemetry metrics and while I find some of them very useful (like the cache hits, number of queries, workers, etc) I am struggling with others. Here are some notes I took:

missing:
 - light client queries and latencies for queries, verification, etc.
 - there is `Number of WriteAcknowledgement relayed` but not `Number of SendPacket relayed`. Should probably fix wording as those are events, hermes acts on events but relays messages, e.g. on `SendPacket` event it relays a `MsgRecvPacket` message, on `WriteAcknowledgement` event it relays `MsgAcknowledgement`.
 - there is a `Number of SendPacket relayed through ClearPendingPackets` but not `Number of WriteAcknowledgement relayed through ClearPendingPackets`. Or is this supposed to count only once per packet lifetime? (i.e. potentially two message). Could this be clarified?
 - number of update client messages

confusing:
 - relayed per channel only shows with `tx_confirmation = true`, help should mention this. But since almost no one runs with this set, I am wondering if we should record the times we send a msg successfully (even if we don't have a confirmation that it was not a duplicate or deliverTx has not failed). Not sure how it will work with retries.
 - `tx_latency_submitted_bucket` always show `le="+Inf"}`. I see the comments below but not clear why we don't have some buckets. We should be able to configure something like `10, 50, 100, 200, 500, 1000, 2000, Inf` no? At least explain there are no buckets or remove it altogether.
      ```
      // TODO: Once quantile sketches are supported, replace histograms with that.
      // For the moment, disable histogram buckets since no values make sense for all use-cases.
      ```

incorrect (?):
the oldest sequence/ timestamp. I send some packets and start hermes with `clear_on_start = false`. Hermes will not relay those packets and can be confirmed with `query packet pending` CLI. However the metrics show:
```
# HELP oldest_sequence The sequence number of the oldest pending SendPacket. If this value is 0, it means there are no pending SendPacket
# TYPE oldest_sequence gauge
oldest_sequence{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 0
# HELP oldest_timestamp The timestamp of the oldest sequence number in seconds
# TYPE oldest_timestamp gauge
oldest_timestamp{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 0
``` 
If hermes relays some packets the metrics above will flicker for a bit with the current sequence number/ timestamp and then go back to 0.

## Version
master

## Steps to Reproduce
- run hermes, send packets, check metrics
- send packets, start hermes, check metrics
- send packets, relay some MsgRecvPacket (use tx raw), send more packets, start hermes
- try with different configuration parameters

## Acceptance Criteria
Better documentation on metrics, what they mean exactly, how they are counted (especially the number of messages, events, etc.). Fix an of the metric inaccuracies. Provide metrics for light client operations as we suspect them to be heavy weights and one of the places we can try improvements. 

____

## For Admin Use

- [x] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate milestone (priority) applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by ljoss17** [comment]
Ok, I suggest that the adding the new metrics to monitor where the relayer spends time will be done in a separate PR to make it easier to review.

I will work on the other changes and link this issue to the PR once it's ready.

> btw, where is "ClearPendingPackets" coming from? And what would be the name for the realtime/ live event mechanism?

I am not sure I understand the question. In the relayer code, `relayer/src/worker/packet.rs` line 107, it seems to be when a `ClearPendingPackets` command is received. Or in the Excalidraw https://app.excalidraw.com/l/4XqkU6POmGI/9X8Mv47sKuD it should be the `Worker::ClearPendingPackets`

> Not necessarily. If some MsgRecvPacket-s have been sent but not the corresponding MsgAcknowledgment-s. For example, i start with 49 pending packets, and MsgRecvPacket sent for the first 30. I expect when I start the relayer, once it's done, to see 19 MsgRecvPacket sent (plus the client updates) and 49 MsgAcknowledgment-s. The metrics show:

I am not sure to understand the example for the `send_packet_count` and `write_acknowledgment_count`.
When Hermes sees any `SendPacket` event, it will increment `send_packet_count`, and additionally the `cleared_count` will show how many of these `SendPacket` events have been observed while relaying messages when the `ClearPendingPackets` command is received. So if we look only at the example where we have 20 pending `MsgRecvPacket`:

* `hermes tx ft-transfer .... --number-msgs 20`

Then start Hermes with the `clear_on_start = true` option, at this point the metrics will be as following:
```
cleared_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
acknowledgement_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
```
As 20 `SendPacket` events have been observed, and of these 20, 20 where observed when Hermes received the `ClearPendingPackets` command. If we now send 15 additional messages while Hermes is running:

* `hermes tx ft-transfer .... --number-msgs 15`

Then the metrics will be:
```
cleared_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 35
acknowledgement_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 35
```
Since the total number of `SendPacket` events seen is 35, but only the first 20 where observed during the `ClearPendingPackets`.
Regarding the Acknowledgments, at the moment there is a potential mismatch. If there are 15 pending `MsgAcknowledgment`, we start Hermes and send an additional 20 messages, the metrics will be:
```
cleared_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 0
send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
acknowledgement_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 35
```
I suggest to add an additional metric, `cleared_acknowledgment_count`, which would count only the `WriteAcknowledgment` observed during the `ClearPendingPackets`. So the previous example would yield the following metrics:
```
cleared_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 0
send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
acknowledgement_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
cleared_acknowledgment_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 15
```

But if we take the case where `clear_on_start = false` and run the following:
1. `hermes tx ft-transfer .... --number-msgs 20`
2. `hermes start`
3. `hermes tx ft-transfer .... --number-msgs 15`

Then the pending packets will not be taken into account as Hermes has not seen the first 20 `SendPacket` and `WriteAcknowledgment` event, thus the metrics will be:
```
cleared_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 0
send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 15
acknowledgement_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 15
```

**Comment by ancazamfir** [comment]
> I am not sure I understand the question. In the relayer code, `relayer/src/worker/packet.rs` line 107, it seems to be when a `ClearPendingPackets` command is received. Or in the Excalidraw https://app.excalidraw.com/l/4XqkU6POmGI/9X8Mv47sKuD it should be the `Worker::ClearPendingPackets`

Do we need to put the name used in the code here? I was asking because it looks out of place.

**Comment by ljoss17** [comment]
I see, I put the ClearPendingPackets as it was in the Excalidraw, which seemed to be the name representing the event. But I can put something else if there is a more appropriate name.

**Comment by ancazamfir** [comment]
> I am not sure to understand the example for the `send_packet_count` and `write_acknowledgment_count`...

I think we say the same thing :) Can we make the count names a bit more consistent:
```
cleared_send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 0
cleared_acknowledgment_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 15
total_send_packet_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
total_acknowledgement_count{chain="ibc-1",channel="channel-0",counterparty="ibc-0",port="transfer"} 20
```
Also the help strings.

**Comment by ljoss17** [comment]
Yes good idea, I'll apply the changes to the PR

**Comment by ancazamfir** [comment]
> I see, I put the ClearPendingPackets as it was in the Excalidraw, which seemed to be the name representing the event. But I can put something else if there is a more appropriate name.

Can we have a full template doc of the exact metrics and help strings? This would help with the review of user UX.

imho we shouldn't assume the user knows the internal architecture and code. We can just say something like:
`# HELP cleared_count number of SendPacket relayed during the initial and periodic clearing.`
The user should know about this functionality. We can add more to the help to say they show only if they are enabled in the configuration.

**Comment by ancazamfir** [comment]
> Ok, I suggest that the adding the new metrics to monitor where the relayer spends time will be done in a separate PR to make it easier to review.

Sure!

**Comment by ancazamfir** [comment]
One more question, is there a way to not show the `ibc_..._packets` counters if confirmation is not set? Or if they are 0?

**Comment by ancazamfir** [comment]
> Can we have a full template doc of the exact metrics and help strings? This would help with the review of user UX.

Actually it's fine, i see they are in the guide.

**Comment by ljoss17** [comment]
Everything except for the new metrics to monitor where the relayer spends time have been added to the PR linked to this issue

**Comment by ancazamfir** [comment]
> Everything except for the new metrics to monitor where the relayer spends time have been added to the PR linked to this issue

thanks @ljoss17 ! i will look at it and ask more questions there.

**Comment by ancazamfir** [comment]
> > missing:
> > 
> > * light client queries and latencies for queries, verification, etc.
> > * number of update client messages
> 
> As there have not been any requests for these metrics, what do you think about having a separate issue as a strech goal for these ?

Ok, I think what I ask for doesn't make a lot of sense wrt to the latency. It would be nice to have counters for how many times we get headers (something in the same category with the queries counters) and how many `MsgUpdateClient` messages we sent (in same group with `ibc_..._packets` counters).

**Comment by ljoss17** [comment]
> > > missing:
> > > 
> > > * light client queries and latencies for queries, verification, etc.
> > > * number of update client messages
> > 
> > 
> > As there have not been any requests for these metrics, what do you think about having a separate issue as a strech goal for these ?
> 
> Ok, I think what I ask for doesn't make a lot of sense wrt to the latency. It would be nice to have counters for how many times we get headers (something in the same category with the queries counters) and how many `MsgUpdateClient` messages we sent (in same group with `ibc_..._packets` counters).

Are the following two metrics the ones you expect ?

* A counter for how many LIghtBlocks we query, which are used to create the IBC headers.
* A counter for how many MsgUpdateClient messages we send.

**Comment by ancazamfir** [comment]
> Are the following two metrics the ones you expect ?
> 
> * A counter for how many LIghtBlocks we query, which are used to create the IBC headers.
> * A counter for how many MsgUpdateClient messages we send.

Let's start with the last one and you can open a new issue/PR for it. In the meantime i will try to find time to do profiling. Thanks!
